File: utils/sentence-splitter.ts
const ABBREVIATIONS = new Set([
  'mr', 'mrs', 'ms', 'dr', 'prof', 'sr', 'jr', 'st', 'ave', 'blvd',
  'gen', 'gov', 'sgt', 'cpl', 'pvt', 'capt', 'lt', 'col', 'maj',
  'rev', 'hon', 'pres', 'dept', 'univ', 'assn', 'bros', 'inc', 'ltd',
  'co', 'corp', 'vs', 'est', 'vol', 'fig', 'eq', 'approx',
  'i.e', 'e.g', 'etc', 'al', 'cf',
])

export function splitSentences(text: string): string[] {
  const sentences: string[] = []
  let current = ''

  const tokens = text.split(/(\s+)/)

  for (const token of tokens) {
    current += token

    // Check if token ends with sentence-ending punctuation
    const match = token.match(/^(.+?)([.!?]+)$/)
    if (!match) continue

    const word = match[1].toLowerCase().replace(/[^a-z.]/g, '')

    // Skip if it's a known abbreviation
    if (ABBREVIATIONS.has(word)) continue

    // Skip single letters followed by period (initials like "J.")
    if (word.length === 1) continue

    // Skip decimal numbers (e.g., "3.14")
    if (/\d$/.test(match[1]) && match[2] === '.') continue

    const trimmed = current.trim()
    if (trimmed) {
      sentences.push(trimmed)
      current = ''
    }
  }

  const remaining = current.trim()
  if (remaining) sentences.push(remaining)

  return sentences
}

File: orchestrator/sentence_splitter.py
import re

ABBREVIATIONS = {
    "mr", "mrs", "ms", "dr", "prof", "sr", "jr", "st", "ave", "blvd",
    "gen", "gov", "sgt", "cpl", "pvt", "capt", "lt", "col", "maj",
    "rev", "hon", "pres", "dept", "univ", "assn", "bros", "inc", "ltd",
    "co", "corp", "vs", "est", "vol", "fig", "eq", "approx",
    "i.e", "e.g", "etc", "al", "cf",
}

IMAGE_MARKER_RE = re.compile(r"^\[image:\d+\]$")


def split_sentences(text: str) -> list[str]:
    sentences: list[str] = []

    # Split on image markers first — they become standalone segments
    parts = re.split(r"(\[image:\d+\])", text)

    for part in parts:
        stripped = part.strip()
        if not stripped:
            continue
        if IMAGE_MARKER_RE.match(stripped):
            sentences.append(stripped)
            continue
        # Split prose into sentences
        sentences.extend(_split_prose(stripped))

    return sentences


def _split_prose(text: str) -> list[str]:
    sentences: list[str] = []
    current = ""

    tokens = re.split(r"(\s+)", text)

    for token in tokens:
        current += token

        match = re.match(r"^(.+?)([.!?]+)$", token)
        if not match:
            continue

        word = re.sub(r"[^a-z.]", "", match.group(1).lower())

        if word in ABBREVIATIONS:
            continue

        if len(word) == 1 and word.isalpha():
            continue

        trimmed = current.strip()
        if trimmed:
            sentences.append(trimmed)
            current = ""

    remaining = current.strip()
    if remaining:
        sentences.append(remaining)

    return sentences
